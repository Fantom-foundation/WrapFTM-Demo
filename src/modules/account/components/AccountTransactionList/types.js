/**
 * AccountTransaction object
 * @typedef {Object} AccountTransaction
 * @property {string} hash
 * @property {string} from
 * @property {string} to
 * @property {string} value
 * @property {string|null} status
 * @property {{timestamp: string}} block
 */
