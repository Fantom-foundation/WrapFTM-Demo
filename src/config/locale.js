import { storeToRefs } from 'pinia';
import { useAppStore } from '@/modules/app/store/store.js';
import { clone, useFormatters } from 'fantom-vue3-components';
import { appConfig } from '@/config/app-config.js';
import { i18n } from '@/config/i18n.js';
import { translations } from 'fantom-vue3-components/src/mixins/translations.js';
import { useLocale } from 'fantom-vue3-components/src/composables/useLocale/useLocale.js';
import { Modules } from '@/utils/Modules/Modules.js';

export async function setUpLocale() {
    const { lang, rtlDirection } = storeToRefs(useAppStore());
    const { locale } = useLocale();
    const { formatters } = useFormatters();

    locale.setup({
        locales: clone(appConfig.locales),
        languageRef: lang,
        rtlDirectionRef: rtlDirection,
        defaultLanguageCode: appConfig.defaultLanguageCode,
        i18n,
        translations,
        formatters,
    });

    await locale.setLocale();

    locale.setTextDirection();
}

export const messageImports = {
    async en() {
        return await Modules.getTranslations('en');
    },
    async cs() {
        return await Modules.getTranslations('cs');
    },
};
