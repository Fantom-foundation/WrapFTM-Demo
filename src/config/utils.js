export function envToBool(value) {
    return value === 'true' || value === '1';
}
