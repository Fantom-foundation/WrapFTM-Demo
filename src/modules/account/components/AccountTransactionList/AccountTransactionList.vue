<script setup>
import './types.js';
import { useI18n } from 'vue-i18n';
import { useFormatters } from 'fantom-vue3-components';
import { FDataGrid, FLink } from 'fantom-vue3-components';
import { ref } from 'vue';
import { bFromWei, toInt } from '@/utils/big-number/big-number.js';
import { addressesMatch } from '@/utils/account/account.js';
import { appConfig } from '@/config/app-config.js';
import { exposeMethods } from 'fantom-vue3-components/src/utils/exposeMethods/exposeMethods.js';

const t = useI18n().t;

const { formatters } = useFormatters();

const props = defineProps({
    /** @type {AccountTransaction[]} */
    transactions: {
        type: [Array, Promise],
        default() {
            return [];
        },
        required: true,
    },
    /** Account address */
    address: {
        type: String,
        default: '',
        required: true,
    },
});

const dataGrid = ref(null);
const columns = ref([
    {
        name: 'status',
        label: t('account.transactionList.status'),
        width: '120px',
        formatter(value) {
            return value !== null ? toInt(value) : null;
        },
    },
    {
        name: 'timestamp',
        label: t('account.transactionList.time'),
        itemProp: 'block.timestamp',
        width: '200px',
        formatter(value) {
            return formatters.dateTime(new Date(toInt(value) * 1000), 'shortDatetime');
        },
    },
    {
        name: 'address',
        label: t('account.transactionList.address'),
        width: '280px',
        formatter(value, item) {
            return getRightAddress(item);
        },
    },
    {
        name: 'amount',
        itemProp: 'value',
        label: `${t('account.transactionList.amount')} (FTM)`,
        width: '200px',
        css: { textAlign: 'end' },
        formatter(value, item) {
            let sign = '';

            if (value !== '0x0') {
                if (addressesMatch(props.address, item.to)) {
                    sign = '+';
                } else {
                    sign = '-';
                }
            }

            return {
                amount:
                    value !== '0x0'
                        ? sign +
                          formatters.number(
                              bFromWei(value),
                              appConfig.formats.ftmFractionDigits,
                              'twoFractionDigits'
                          )
                        : '0',
                highlight: sign === '+',
            };
        },
    },
]);

function getStatusLabel(status) {
    let label = t('account.transactionList.pending');

    if (status === 0) {
        label = t('account.transactionList.fail');
    } else if (status === 1) {
        label = t('account.transactionList.success');
    }

    return label;
}

function getStatusCssClass(status) {
    let clas = [];

    if (status === 0) {
        clas.push('accounttransactionlist_status-fail');
    } else if (status === 1) {
        clas.push('accounttransactionlist_status-success');
    }

    return clas;
}

function getRightAddress(transaction) {
    return !addressesMatch(props.address, transaction.to)
        ? transaction.to
        : transaction.from;
}

defineExpose(exposeMethods(dataGrid, ['resetPagination']));
</script>

<template>
    <FDataGrid
        ref="dataGrid"
        :columns="columns"
        :items="transactions"
        strategy="remote"
        no-header
        infinite-scroll
        infinite-scroll-root-margin="200px 0px"
        sticky-head
        class="accounttransactionlist"
    >
        <template #column-status="{ value: status }">
            <span
                class="accounttransactionlist_status"
                :class="getStatusCssClass(status)"
            >
                {{ getStatusLabel(status) }}
            </span>
        </template>

        <template #column-timestamp="{ value: date, item: transaction }">
            <FLink
                type="transaction"
                :hash="transaction.hash"
                :text="date"
                target="_blank"
            />
        </template>

        <template #column-address="{ value: address }">
            <FLink type="address" :text="address || ''" use-ellipsis target="_blank" />
        </template>

        <template #column-amount="{ value: { amount, highlight } }">
            <span
                class="accounttransactionlist_amount"
                :class="{
                    'accounttransactionlist_amount-highligted': highlight,
                }"
            >
                {{ amount }}
            </span>
        </template>
    </FDataGrid>
</template>

<style lang="scss">
.accounttransactionlist {
    &_status {
        font-weight: bold;

        &-fail {
            color: var(--f-color-red-5);
        }

        &-success {
            color: var(--f-color-green-5);
        }
    }

    &_amount {
        font-weight: bold;

        &-highligted {
            color: var(--f-color-green-5);
        }
    }
}
</style>
