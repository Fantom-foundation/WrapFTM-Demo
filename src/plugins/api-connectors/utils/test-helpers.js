export class MyApiConnectorMock {
    __errors__ = {};
    __results__ = {};

    async processQuery() {
        return new Promise((resolve, reject) => {
            // console.log(this.__errors__.processQuery);
            if (this.__errors__.processQuery !== undefined) {
                reject(this.__errors__.processQuery);
            } else if (this.__results__.processQuery !== undefined) {
                resolve(this.__results__.processQuery);
            } else {
                resolve('result');
            }
        });
    }

    async processMutation() {
        return new Promise((resolve, reject) => {
            // console.log(this.__errors__.processQuery);
            if (this.__errors__.processMutation !== undefined) {
                reject(this.__errors__.processMutation);
            } else if (this.__results__.processMutation !== undefined) {
                resolve(this.__results__.processMutation);
            } else {
                resolve('result');
            }
        });
    }

    processResult(result) {
        return result;
    }
}
