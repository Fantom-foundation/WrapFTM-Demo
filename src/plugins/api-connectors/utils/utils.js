import { isProxy, isReactive, isRef, toRaw, unref } from 'vue';

/**
 * @param {function[]} functions
 * @param args
 */
export function callFunctions(functions, ...args) {
    functions.forEach((fn) => {
        if (typeof fn === 'function') {
            fn(...args);
        }
    });
}

export function deepToRaw(sourceObj) {
    const objectIterator = (input) => {
        if (Array.isArray(input)) {
            return input.map((item) => objectIterator(item));
        }
        if (isRef(input)) {
            return objectIterator(unref(input));
        }
        if (isReactive(input) || isProxy(input)) {
            return objectIterator(toRaw(input));
        }
        if (input && typeof input === 'object') {
            return Object.getOwnPropertyNames(input).reduce((acc, key) => {
                acc[key] = objectIterator(input[key]);
                return acc;
            }, {});
        }
        return input;
    };

    return objectIterator(sourceObj);
}
