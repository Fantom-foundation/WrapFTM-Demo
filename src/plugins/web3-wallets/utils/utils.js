import { toInt } from '@/utils/big-number/big-number.js';
import { CHAIN_NAMES } from '@/plugins/web3-wallets/utils/chain-names.js';

export function getChainNameById(chainId) {
    const chainIdI = toInt(chainId);

    return CHAIN_NAMES[chainIdI] || '';
}
