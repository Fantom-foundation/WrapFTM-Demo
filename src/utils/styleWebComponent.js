function applyStyles(elem, styles) {
    Object.keys(styles).forEach((prop) => {
        elem.style[prop] = styles[prop];
    });
}

function findWebComponent(wcSelector) {
    const wcs = wcSelector.split('>');
    let wc = null;

    wcs.forEach((wcS) => {
        wc = (wc === null ? document : wc?.shadowRoot)?.querySelector(wcS);
    });

    return wc;
}

/**
 * @param {string} wcSelector
 * @param {Object<string, Object>} styles
 */
export function styleWebComponent(wcSelector, styles) {
    const wc = findWebComponent(wcSelector);

    if (!wc) {
        throw new Error(`Can't find web component '${wcSelector}'`);
    }

    if (!wc.shadowRoot) {
        throw new Error(`'${wcSelector}' is not a web component`);
    }

    Object.keys(styles).forEach((selector) => {
        if (selector === 'root') {
            applyStyles(wc, styles[selector]);
        } else {
            const elems = wc.shadowRoot.querySelectorAll(selector);

            if (elems.length > 0) {
                elems.forEach((elem) => {
                    applyStyles(elem, styles[selector]);
                });
            } else {
                console.warn(`Can't find element(s) '${selector}'`);
            }
        }
    });
}
