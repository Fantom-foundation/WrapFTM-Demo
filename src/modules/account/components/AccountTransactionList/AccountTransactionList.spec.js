import { mount } from '@vue/test-utils';
import { destroyWrapper } from 'fantom-vue3-components/src/test/utils.js';
import AccountTransactionList from './AccountTransactionList.vue';
import {
    TEST_ACCOUNT_ADDRESS,
    TEST_ACCOUNT_ADDRESS2,
} from '@/plugins/web3-wallets/test-helpers.js';
import { setUpTestLocales } from '@/config/test/locale.js';
import { nextTick } from 'vue';

let wrapper = null;

function TRANSACTIONS() {
    return [
        {
            hash: '0x39852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2',
            from: TEST_ACCOUNT_ADDRESS,
            to: TEST_ACCOUNT_ADDRESS2,
            value: '0x0',
            status: '0x1',
            block: {
                timestamp: '0x63289dba',
            },
        },
        {
            hash: '0x40852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2',
            from: TEST_ACCOUNT_ADDRESS,
            to: TEST_ACCOUNT_ADDRESS2,
            value: '0x6B14BD1E6EEA00000',
            status: '0x0',
            block: {
                timestamp: '0x631f6aa6',
            },
        },
        {
            hash: '0x80852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2',
            from: TEST_ACCOUNT_ADDRESS2,
            to: TEST_ACCOUNT_ADDRESS,
            value: '0xde0b6b3a7640000',
            status: null,
            block: {
                timestamp: '0x631f6aa6',
            },
        },
    ];
}

function createWrapper(
    options = {
        props: {
            transactions: TRANSACTIONS(),
            address: TEST_ACCOUNT_ADDRESS,
        },
    }
) {
    return mount(AccountTransactionList, options);
}

beforeAll(() => {
    setUpTestLocales();
});

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('AccountTransactionList', () => {
    it('should display transaction status', () => {
        wrapper = createWrapper();
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        expect(gridRows[0].text()).toContain('Success');
        expect(gridRows[1].text()).toContain('Fail');
        expect(gridRows[2].text()).toContain('Pending');
    });

    it('should display transaction timestamp', () => {
        wrapper = createWrapper();
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        expect(gridRows[0].text()).toContain('9/19/2022, 4:50 PM');
        expect(gridRows[1].text()).toContain('9/12/2022, 5:21 PM');
        expect(gridRows[2].text()).toContain('9/12/2022, 5:21 PM');
    });

    it('should display `to` address if it differs from given account address', () => {
        wrapper = createWrapper();
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        expect(gridRows[0].text()).toContain(TEST_ACCOUNT_ADDRESS2);
    });

    it('should display `from` address if it differs from given account address', () => {
        wrapper = createWrapper();
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        expect(gridRows[2].text()).toContain(TEST_ACCOUNT_ADDRESS2);
    });

    it('should display transferred amount', () => {
        wrapper = createWrapper({
            props: {
                transactions: TRANSACTIONS(),
                address: TEST_ACCOUNT_ADDRESS,
            },
        });
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        expect(gridRows[0].find('td:last-child').text()).toBe('0');
        expect(gridRows[1].find('td:last-child').text()).toBe('-123.46');
        expect(gridRows[2].find('td:last-child').text()).toBe('+1.00');
    });

    it('should display link to transaction detail in a fantom explorer', () => {
        wrapper = createWrapper({
            props: {
                transactions: TRANSACTIONS(),
                address: TEST_ACCOUNT_ADDRESS,
            },
        });
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        expect(gridRows[0].find('a').attributes('href')).toBe(
            'https://ftmscan.com/tx/0x39852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2'
        );
    });

    it('should display link to address detail in a fantom explorer', () => {
        wrapper = createWrapper({
            props: {
                transactions: TRANSACTIONS(),
                address: TEST_ACCOUNT_ADDRESS,
            },
        });
        const gridRows = wrapper.findByTestCode('fdatagrid-row');

        const links = gridRows[0].findAll('a');
        expect(links[1].attributes('href')).toBe(
            `https://ftmscan.com/address/${TEST_ACCOUNT_ADDRESS2}`
        );
    });

    it('should expose FDataGrid methods', () => {
        wrapper = createWrapper();

        expect(wrapper.vm.$parent.$refs.VTU_COMPONENT.resetPagination).toBeDefined();
    });

    it('should match snapshot', async () => {
        wrapper = createWrapper({
            props: {
                transactions: TRANSACTIONS(),
                address: TEST_ACCOUNT_ADDRESS,
                addressUrl: 'https://ftmscan.com/address/',
            },
        });

        await nextTick();
        // remove dynamic attributes
        wrapper.removeAttributes('id');
        wrapper.removeAttributes('data-fdg-id');

        expect(wrapper.element).toMatchSnapshot();
    });
});
