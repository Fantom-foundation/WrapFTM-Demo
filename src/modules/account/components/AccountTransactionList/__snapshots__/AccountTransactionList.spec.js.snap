// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountTransactionList > should match snapshot 1`] = `
<div
  addressurl="https://ftmscan.com/address/"
  class="fdatagrid fdatagrid-stickyheader fdatagrid-infinitescrollon fdatagrid-noheader accounttransactionlist"
>
  <!--v-if-->
  <div
    class="fdatagrid_table"
  >
    <table
      role="grid"
    >
      <!--v-if-->
      <thead
        class="fdatagrid_thead"
      >
        <tr>
          
          <th
            class="_c0  fdatagrid_firstcell"
          >
            <div
              class="fdatagrid_thead_cellin"
            >
              Status
            </div>
          </th>
          <th
            class="_c1"
          >
            <div
              class="fdatagrid_thead_cellin"
            >
              Time
            </div>
          </th>
          <th
            class="_c2"
          >
            <div
              class="fdatagrid_thead_cellin"
            >
              Address
            </div>
          </th>
          <th
            class="_c3  fdatagrid_lastcell"
          >
            <div
              class="fdatagrid_thead_cellin"
            >
              Amount (FTM)
            </div>
          </th>
          
        </tr>
      </thead>
      <tbody
        class="finfinitescroll"
      >
        <tr
          class="fintersectionobserver"
          style="display: none;"
        >
          
          
          <td
            colspan="4"
          >
            
            <div
              class="tea-center"
            >
              <span
                class="fdotsloader"
              >
                <span />
                <span />
                <span />
              </span>
            </div>
            
          </td>
          
          
        </tr>
        
        
        
        <tr
          aria-rowindex="1"
          class=""
          data-testcode="fdatagrid-row"
        >
          
          <td
            class="_c0  fdatagrid_firstcell"
            data-label="Status"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <span
                class="accounttransactionlist_status accounttransactionlist_status-success"
              >
                Success
              </span>
              
              
            </div>
          </td>
          <td
            class="_c1"
            data-label="Time"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <a
                class="flink"
                data-type="transaction"
                href="https://ftmscan.com/tx/0x39852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2"
                target="_blank"
              >
                
                
                9/19/2022, 4:50 PM
                
                
              </a>
              
              
            </div>
          </td>
          <td
            class="_c2"
            data-label="Address"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <a
                class="flink"
                data-type="address"
                href="https://ftmscan.com/address/0xb8CE9ab6943e0eCED004cDe8e3bBed6568B2Fa01"
                target="_blank"
              >
                
                <span
                  class="fellipsis overflow-middle"
                >
                  <span
                    class="fellipsis_prefix"
                  >
                    
                    
                  </span>
                  
                  <span
                    class="left-chars"
                  >
                    0xb8CE9ab6943e0eCED004cDe8e3bBed6568B
                  </span>
                  <span
                    class="right-chars"
                  >
                    2Fa01
                  </span>
                  
                  <span
                    class="fellipsis_suffix"
                  >
                    
                    
                  </span>
                </span>
                
              </a>
              
              
            </div>
          </td>
          <td
            class="_c3  fdatagrid_lastcell"
            data-label="Amount (FTM)"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <span
                class="accounttransactionlist_amount"
              >
                0
              </span>
              
              
            </div>
          </td>
          
        </tr>
        
        
        
        
        <tr
          aria-rowindex="2"
          class=""
          data-testcode="fdatagrid-row"
        >
          
          <td
            class="_c0  fdatagrid_firstcell"
            data-label="Status"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <span
                class="accounttransactionlist_status accounttransactionlist_status-fail"
              >
                Fail
              </span>
              
              
            </div>
          </td>
          <td
            class="_c1"
            data-label="Time"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <a
                class="flink"
                data-type="transaction"
                href="https://ftmscan.com/tx/0x40852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2"
                target="_blank"
              >
                
                
                9/12/2022, 5:21 PM
                
                
              </a>
              
              
            </div>
          </td>
          <td
            class="_c2"
            data-label="Address"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <a
                class="flink"
                data-type="address"
                href="https://ftmscan.com/address/0xb8CE9ab6943e0eCED004cDe8e3bBed6568B2Fa01"
                target="_blank"
              >
                
                <span
                  class="fellipsis overflow-middle"
                >
                  <span
                    class="fellipsis_prefix"
                  >
                    
                    
                  </span>
                  
                  <span
                    class="left-chars"
                  >
                    0xb8CE9ab6943e0eCED004cDe8e3bBed6568B
                  </span>
                  <span
                    class="right-chars"
                  >
                    2Fa01
                  </span>
                  
                  <span
                    class="fellipsis_suffix"
                  >
                    
                    
                  </span>
                </span>
                
              </a>
              
              
            </div>
          </td>
          <td
            class="_c3  fdatagrid_lastcell"
            data-label="Amount (FTM)"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <span
                class="accounttransactionlist_amount"
              >
                -123.46
              </span>
              
              
            </div>
          </td>
          
        </tr>
        
        
        
        
        <tr
          aria-rowindex="3"
          class=""
          data-testcode="fdatagrid-row"
        >
          
          <td
            class="_c0  fdatagrid_firstcell"
            data-label="Status"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <span
                class="accounttransactionlist_status"
              >
                Pending
              </span>
              
              
            </div>
          </td>
          <td
            class="_c1"
            data-label="Time"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <a
                class="flink"
                data-type="transaction"
                href="https://ftmscan.com/tx/0x80852eca227233345f541691124ad91d30800534035dea0e67a57627652acbb2"
                target="_blank"
              >
                
                
                9/12/2022, 5:21 PM
                
                
              </a>
              
              
            </div>
          </td>
          <td
            class="_c2"
            data-label="Address"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <a
                class="flink"
                data-type="address"
                href="https://ftmscan.com/address/0xb8CE9ab6943e0eCED004cDe8e3bBed6568B2Fa01"
                target="_blank"
              >
                
                <span
                  class="fellipsis overflow-middle"
                >
                  <span
                    class="fellipsis_prefix"
                  >
                    
                    
                  </span>
                  
                  <span
                    class="left-chars"
                  >
                    0xb8CE9ab6943e0eCED004cDe8e3bBed6568B
                  </span>
                  <span
                    class="right-chars"
                  >
                    2Fa01
                  </span>
                  
                  <span
                    class="fellipsis_suffix"
                  >
                    
                    
                  </span>
                </span>
                
              </a>
              
              
            </div>
          </td>
          <td
            class="_c3  fdatagrid_lastcell"
            data-label="Amount (FTM)"
            tabindex="-1"
          >
            <div
              class="fdatagrid_cellin"
            >
              
              <!-- @slot Dynamic slot for cells -->
              
              <span
                class="accounttransactionlist_amount accounttransactionlist_amount-highligted"
              >
                +1.00
              </span>
              
              
            </div>
          </td>
          
        </tr>
        
        
        
        
        
        <tr
          class="fintersectionobserver"
          style="display: none;"
        >
          
          
          <td
            colspan="4"
          >
            
            <div
              class="tea-center"
            >
              <span
                class="fdotsloader"
              >
                <span />
                <span />
                <span />
              </span>
            </div>
            
          </td>
          
          
        </tr>
        <div
          class="fpagination"
          hidden=""
          style="display: none;"
        >
          <ul
            aria-label="Pagination navigation"
            class="fpagination_body"
          >
            <li
              class="fpagination_item"
              data-item-type="first"
            >
              <!--
                    @slot Slot for buttons
                    @binding {string} label
                    @binding {string} title
                    @binding {boolean} disabled
                    @binding {('first'|'prev'|'next'|'last')} type
                -->
              
              <button
                class="fbutton btn btn-tertiary btn-round btn-samesize"
                disabled=""
                title="Go to first page"
                type="button"
              >
                <span
                  class="fbutton_label"
                >
                  
                  «
                  
                </span>
              </button>
              
            </li>
            <li
              class="fpagination_item"
              data-item-type="prev"
            >
              
              <button
                class="fbutton btn btn-tertiary btn-round btn-samesize"
                disabled=""
                title="Go to previous page"
                type="button"
              >
                <span
                  class="fbutton_label"
                >
                  
                  ‹
                  
                </span>
              </button>
              
            </li>
            <li
              class="fpagination_pages"
            >
              <!--
                    @slot Pages info slot
                    @binding {FPaginationState} state
                -->
              
              
              <span>
                -
              </span>
              
              
            </li>
            <li
              class="fpagination_item"
              data-item-type="next"
            >
              
              <button
                class="fbutton btn btn-tertiary btn-round btn-samesize"
                disabled=""
                title="Go to next page"
                type="button"
              >
                <span
                  class="fbutton_label"
                >
                  
                  ›
                  
                </span>
              </button>
              
            </li>
            <li
              class="fpagination_item"
              data-item-type="last"
            >
              
              <button
                class="fbutton btn btn-tertiary btn-round btn-samesize"
                disabled=""
                title="Go to last page"
                type="button"
              >
                <span
                  class="fbutton_label"
                >
                  
                  
                  »
                  
                  
                </span>
              </button>
              
            </li>
          </ul>
        </div>
      </tbody>
      <!--v-if-->
    </table>
    
    <!--v-if-->
    
  </div>
  <div
    class="fdatagrid_footer"
  >
    <!--v-if-->
  </div>
  <div
    style="display: none;"
  />
</div>
`;
